server/jetstream_cluster.go

```diff
@@ -4172,8 +4434,79 @@ func (s *Server) jsClusteredStreamUpdateRequest(ci *ClientInfo, acc *Account, su
 			}
 			rg.Peers = peers
 		} else {
-			// We are deleting nodes here.
-			rg.Peers = rg.Peers[:newCfg.Replicas]
+			// We are deleting nodes here. We want to do our best to preserve the current leader.
+			// We have support now from above that guarantees we are in our own Go routine, so can
+			// ask for stream info from the stream leader to make sure we keep the leader in the new list.
+			var curLeader string
+			if !s.allPeersOffline(rg) {
+				// Need to release js lock.
+				js.mu.Unlock()
```

Comment:
Not a big fan, as you know, but I understand why you do it here. Maybe we should have had a specific lock for the s.sys.replies so that we could do it just for the registration of those handlers, without requiring server lock, but ok.

Reply:
Actually, you don't need to unlock. I created a file called `locksordering.txt` where I list the known ordering, and it says:
```
jetStream -> jsAccount -> Server -> client-> Account
```
So it is ok to get server lock inside js lock.

Reply:
ok great will make that change.

Reply:
Check that, this calls out to other servers and we do not want to hold any locks across that, so will not make that change. But good to know.

Reply:
Not while we wait, I guess. But then you could do the cleanup of replies outside of the msgHandler, or at least after the select in case we did not receive a msg.

Reply:
Yep good call.